src/test/blockmanager_tests.cpp

```diff
@@ -82,4 +83,113 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
     BOOST_CHECK(!AutoFile(OpenBlockFile(new_pos, true)).IsNull());
 }
 
+struct FastPruneTestingSetup : public BasicTestingSetup {
+    FastPruneTestingSetup()
+        : BasicTestingSetup{CBaseChainParams::MAIN, /*extra_args=*/{"-fastprune=1"}} {}
+};
+
+BOOST_FIXTURE_TEST_CASE(blockmanager_flush_block_file, FastPruneTestingSetup)
+{
+    const auto params{CreateChainParams(ArgsManager{}, CBaseChainParams::MAIN)};
+    BlockManager blockman{};
+    CChain chain{};
+
+    // Test blocks with no transactions, not even a coinbase
+    CBlock block1;
+    block1.nVersion = 1;
+    CBlock block2;
+    block2.nVersion = 2;
+    CBlock block3;
+    block3.nVersion = 3;
+
+    // They are 80 bytes header + 1 byte 0x00 for vtx length
+    int TEST_BLOCK_SIZE = 81;
+
+    // Blockstore is empty
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), 0);
+
+    // Write first block
+    FlatFilePos pos1 = blockman.SaveBlockToDisk(block1, 1, chain, *params, nullptr);
+
+    // Write second block
+    FlatFilePos pos2 = blockman.SaveBlockToDisk(block2, 2, chain, *params, nullptr);
+
+    // Two blocks in the file
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), (TEST_BLOCK_SIZE + BLOCK_SERIALIZATION_HEADER_SIZE) * 2);
+
+    // First two blocks are written as expected
+    CBlock read_block;
+    node::ReadBlockFromDisk(read_block, pos1, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 1);
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // Record last modified time of blk00000.dat
+    const fs::path file_path = node::GetBlockPosFilename(pos2);
+    const fs::file_time_type time1 = fs::last_write_time(file_path);
+
+#ifdef WIN32
+    // Ensure that if the file is modified its timestamp is updated,
+    // even on Windows (FAT has 2 second resolution)
+    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+#endif
+
+    // Attempt to save block 3 to original position of block 2
+    blockman.SaveBlockToDisk(block3, 3, chain, *params, &pos2);
+
+    // Still only two blocks in the file
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), (TEST_BLOCK_SIZE + BLOCK_SERIALIZATION_HEADER_SIZE) * 2);
+
+    // Block 2 was not overwritten:
+    //   SaveBlockToDisk() did not call WriteBlockToDisk() because `FlatFilePos* dbp` was non-null
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // blk00000.dat was not modified
+    const fs::file_time_type time2 = fs::last_write_time(file_path);
+    BOOST_CHECK(time1 == time2);
```

Comment:
Well, on non-windows, it _may_ have been modified, but within the same second. This is kind of a minor point (probably don't need to change anything), but if in the future there's a bug such that `blk00000.dat` was modified, this test might not catch it. I verified this by hacking the test such that the file _is_ modified, and about once in 10 or 20 runs, the timestamp didn't change. (This is on Ubuntu.)

But I guess that's okay, because if there is such a bug, it will likely be caught quickly. It's just that I wish this test could be 100 percent solid. The only way I could think of to do that is to read and checksum the file, and compare the checksums instead of the timestamps. (Then you wouldn't need that sleep for WIN32.) It wouldn't be a performance problem, because block files are very small in this test.

Just something to consider, it may not be worth the trouble.

Reply:
> The only way I could think of to do that is to read and checksum the file, and compare the checksums instead of the timestamps. 

Isn't the point that the file contents aren't changing? So checksum wouldn't work because the files will be the same whether it's rewritten or not. The only things that can be different are the filesystem's metadata on the modification time.

Reply:
> The only things that can be different are the filesystem's metadata on the modification time, and these are not covered by a checksum.

It might be the same usefulness actually, I seem to remember that if file contents aren't changed, when you call `fwrite()` or `fflush()`, the modification timestamp won't be updated. I researched different ways to detect a `fflush()` and this unit test was the closest. Most systems flush when `fwrite()` is called, and then a subsequent `fflush()` doesn't do anything at all. The best I could do with this unit test was cover current behavior to ensure that the PR didn't break it.

So actually it might be reasonable to just ditch the timestamp stuff altogether and just ensure that file contents aren't modified.

The better test coverage is in the functional test which will throw on master when bitcoin tries to **open** the read-only file.